File: 0110-balanced-binary-tree/0110-balanced-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        # binary tree = recursive always
        """
        height of the left subtree and right subtree of any node does not differ by more than 1
        """
        return self.height(root) >= 0
        
    def height(self, root):
        if not root: return 0
        left_height, right_height = self.height(root.left), self.height(root.right)
        
        # if subtree differ more than 1
        if left_height < 0 or right_height < 0 or abs(left_height - right_height) > 1:
            return -1
        
        # return the height of this sub tree
        return max(left_height, right_height) + 1
            
        

File: 0110-balanced-binary-tree/0110-balanced-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public boolean isBalanced(TreeNode root) {
        // a height balanced tree has onnly a max depth diff of 1 between subtrees
        return height(root) >= 0;
    }
    
    public int height(TreeNode root) {
        if (root == null) return 0;
        int left_height = height(root.left);
        int right_height = height(root.right);
        
        if (left_height < 0 || right_height < 0 || Math.abs(left_height - right_height) > 1) return -1;
        
        return Math.max(left_height, right_height)+1;
    }
}